src/osd/OSD.cc

```diff
@@ -10146,6 +10146,17 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
     sdata->sdata_op_ordering_lock.Unlock();
     return;    // OSD shutdown, discard.
   }
+  if (item.second.is_future()) {
+    utime_t t = item.second.get_future_time();
+    dout(20) << __func__ << " dequeue future request, pgid: " << item.first
+             << ", delay " << t.to_nsec() - ceph_clock_now().to_nsec() << " nanoseconds." << dendl;
+    sdata->sdata_lock.Lock();
+    sdata->sdata_op_ordering_lock.Unlock();
+    sdata->sdata_cond.WaitUntil(sdata->sdata_lock, t);
```

Comment:
I'm concerned that we might wait too long for the future to become active in some circumstances. If I traced through the code right, `_process` is being called by `ShardedThreadPool::shardedthreadpool_worker`. And it seems to manipulate the heartbeat map upon return from process before looping back around. And furthermore, when the queue is empty, our wait time is bounded by `threadpool_empty_queue_max_wait`.

Reply:
Yes, I think I have lost the heartbeat map reset.  if we use the new solution mentioned above, these problems will gone.